### PR TITLE
Add unbound typevar check

### PR DESCRIPTION
**Pull request by AurelienJaquier**
### Description

This should fix Issue #13061 .

Mypy should now raise an error when it encounters unbound plain TypeVar.
It should not get triggered when the return type contains a TypeVar (e.g. List[T]).
It should not get triggered when a same TypeVar is present in the scope (e.g. inner function returns T, and T is an argument of outer function).

## Test Plan

5 tests were added:

- a plain unbound typevar triggering the new error
- a typevar inner function not triggering the error
- a function returning an iterable of typevars, not triggering the error
- a plain bound typevar, not triggering the error
- nested bound typevars, not triggering the error

We also changed the other functions triggering our error for the tests to pass.

This is our 1st contribution to Mypy. Please, guide us if there is anything we can improve in this PR.

This PR was made with @anilbey

### CODE REVIEW

**Review by ilevkivskyi — APPROVED** [review]
Thanks, looks good! (assuming the tests pass)

**Comment by anilbey** [comment]
Good news! 🎉

File: mypy/checker.py [hunk 1]
```diff
@@ -1062,6 +1065,15 @@ def check_func_def(self, defn: FuncItem, typ: CallableType, name: Optional[str])
 
             self.binder = old_binder
 
+    def check_unbound_return_typevar(self, typ: CallableType) -> None:
+        """Fails when the return typevar is not defined in arguments."""
+        arg_type_visitor = CollectArgTypes()
+        for argtype in typ.arg_types:
+            argtype.accept(arg_type_visitor)
+
+        if typ.ret_type not in arg_type_visitor.arg_types and typ.ret_type in typ.variables:
```
Comment (hauntsaninja) [inline mypy/checker.py:1074]: Could we micro optimise this by first checking `if typ.ret_type in typ.variables` and only then doing the collection of arg types?

File: test-data/unit/check-typevar-unbound.test [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+[case testNestedBoundTypeVar]
+from typing import Callable, Union, TypeVar
+
+T = TypeVar('T')
+
+def f(a: Union[int, T], b: str) -> T:
+    ...
+
+def g(a: Callable[..., T], b: str) -> T:
```
Comment (hauntsaninja) [inline test-data/unit/check-typevar-unbound.test:52]: could you add one more test case with something more nested, say List[Union[Callable[..., Tuple[T]]]]

**Review by hauntsaninja — COMMENTED** [review]
Could you add a test for upper bounded and constrained type vars?

I would also avoid opening PRs to projects that add `type: ignore`, since they likely won't be able to merge the PR until there's a new version of mypy. Besides, fixes are much preferred to adding type-ignores. If we just want to type-ignore most instances of this error, it'd be a sign that we shouldn't add it ;-)

**Comment by hauntsaninja** [comment]
If the typevar is upper bounded or constrained, we could actually suggest using the upper bound or the union of the constraints as the return type. This seems reasonably common, e.g. see the pandas case.

**Comment by ilevkivskyi** [comment]
Btw, this reminds me some people use returning a plain unbound type var with values as a way to emulate unsafe unions. But as many of you know I don't like unsafe unions, so I am fine with breaking such hacks.

**Comment by ilevkivskyi** [comment]
@anilbey @AurelienJaquier it would be great if you could implement some of the suggestions, but if you don't have time or want to do it in a separate PR it is also OK, please let me know.

File: mypy/checker.py (hunk 1 above)
Comment (anilbey) [reply mypy/checker.py:1074]: Good idea!

**Comment by anilbey** [comment]
Thanks @hauntsaninja for the feedback.
@ilevkivskyi I just added the deeper nested tests and micro optimisation. For upper bounded and constrained types, I will need a bit of more time.

**Comment by ilevkivskyi** [comment]
OK, you can make those other changes in a separate PR. I will merge this when the tests pass.

**Comment by anilbey** [comment]
> OK, you can make those other changes in a separate PR. I will merge this when the tests pass.
Alright, sounds good

**Comment by evbo** [comment]
@AurelienJaquier I used to be able to do things like:

```python
def my_func() -> T:

a: MyType = my_func()
```

Now, to avoid `cast` and not rely on `Any` or no typing, I must do:
```python
def my_func(_: Optional[T] = None) -> T:

a: MyType = my_func()
```

Did you consider this use-case? What about upperbounded types? `[T: UpperType]` that's still "unbounded" to you?

**Pull request merged**